### PR TITLE
contains(): Return an array of booleans when array passed

### PR DESCRIPTION
**Pull request by efeichen**
Add logic so an array of booleans would be returned if haystack of type array is passed to the function (w/o been called by another `contains()`).

### CODE REVIEW

File: src/functions.js [hunk 1]
```diff
@@ -324,6 +324,36 @@ var _ = Mavo.Functions = {
...
+     * Search if a group, collection, or primitive contains needle
+	 * @returns Boolean if a haystack of object or primitive is passed
+	 * @returns Array of booleans if a haystack of array is passed
+     */
+    contains: (haystack, needle) => {
+		var ret = false;
+		if ($.type(haystack) === "object") {
+			for (var property in haystack) {
+				ret = _.contains(haystack[property], needle);
+				if ($.type(ret) === "array") {
```
Comment (LeaVerou) [inline src/functions.js]: Prefer `Array.isArray()`, it's more readable and more efficient.

File: src/functions.js [hunk 2]
```diff
@@ -324,6 +324,36 @@ var _ = Mavo.Functions = {
...
+				if (ret) { 
+					return true; 
+				}
+			}
+		}		
+		else if ($.type(haystack) === "array") {
+			ret = [];
+			for (var i = 0; i < haystack.length; i++) {
+				ret.push(_.contains(haystack[i], needle));
+			}
```
Comment (LeaVerou) [inline src/functions.js]: This works well, but using `haystack.map()` would be more readable/elegant.

File: src/functions.js [hunk 3]
```diff
@@ -324,6 +324,36 @@ var _ = Mavo.Functions = {
...
+    contains: (haystack, needle) => {
+		var ret = false;
+		if ($.type(haystack) === "object") {
+			for (var property in haystack) {
+				ret = _.contains(haystack[property], needle);
+				if ($.type(ret) === "array") {
+					ret = Mavo.Functions.or(ret);
+				}
+				if (ret) { 
+					return true;
```
Comment (LeaVerou) [inline src/functions.js]: Why this? Won't the `return ret` in the end suffice?

**Comment by LeaVerou** [comment]
Note that it still doesn't handle the case where `needle` is an array.
I'd repeat my recommendation for looking into `Mavo.Script.binaryOperation()`, which would handle both that case and the array case (that you are currently handling manually).

File: src/functions.js (hunk 3 above)
Comment (efeichen) [reply src/functions.js]: If ret is false it shouldn't be returned and terminate the function yet since there might be a match later in the object.

**Comment by efeichen** [comment]
@LeaVerou If `needle` is an array, how should the haystack be searched? If a haystack of object is passed should it return true if the object contains all items in the array? Or does the object has to have a nested array of the exact same content as the needle array?

File: src/functions.js [hunk 4]
```diff
@@ -324,6 +324,38 @@ var _ = Mavo.Functions = {
...
+						ret = _.contains(haystack[property], needle);
+						if (Array.isArray(ret)) {
+							ret = Mavo.Functions.or(ret);
+						}
+						if (ret) { 
+							return true;
+						}
+					}
+				}
+				else if (type === "array") {
```
Comment (LeaVerou) [inline src/functions.js]: Have you tested without this else if? `binaryOperation()` should handle this case...

File: src/functions.js [hunk 5]
```diff
@@ -324,6 +324,38 @@ var _ = Mavo.Functions = {
 
 	len: text => str(text).length,
 
+	/**
+     * Search if a group, collection, or primitive contains needle
+	 * @returns Boolean if a haystack of object or primitive is passed
+	 * @returns Array of booleans if a haystack of array is passed
+     */
+    contains: (haystack, needle) => {
+		var type = $.type(haystack);
```
Comment (LeaVerou) [inline src/functions.js]: This should be inside the scalar function.

File: src/functions.js (hunk 4 above)
Comment (efeichen) [reply src/functions.js]: Yes! I deleted that block and nothing broke.

File: src/functions.js [hunk 6]
```diff
@@ -324,10 +324,48 @@ var _ = Mavo.Functions = {
...
+	/**
+     * Search if a group, collection, or primitive contains needle
+	 * @returns Boolean if a haystack of object or primitive is passed
+	 * @returns Array of booleans if a haystack of array is passed
+     */
+    contains: (haystack, needle) => {
+		var ret = Mavo.Script.binaryOperation(haystack, needle, {
+			scalar: (haystack, needle) => {
+				var type = $.type(haystack);
+				if (type === "object") {
```
Comment (LeaVerou) [inline src/functions.js]: We don’t need the `type` variable here anymore, we're only using this once. The whole point of the variable was to cache the results so we don't have to run the function twice.

File: src/functions.js [hunk 7]
```diff
@@ -324,10 +324,48 @@ var _ = Mavo.Functions = {
...
+					for (var property in haystack) {
```
Comment (LeaVerou) [inline src/functions.js]: We don’t need the type variable here anymore, we're only using this once. The whole point of the variable was to cache the results so we don't have to run the function twice.

File: src/functions.js [hunk 8]
```diff
@@ -324,10 +324,48 @@ var _ = Mavo.Functions = {
...
+				}
+				else {
+					return _.search(haystack, needle) >= 0;
+				}
+				return ret;
+			},
+			identity: null
+		});
+		// if result is an empty array, return false
+		if (ret.length === 0) {
```
Comment (LeaVerou) [inline src/functions.js]: Very minor nitpick: Blank lines can often improve readability (provided we don’t overuse them!). My usual rule of thumb is I use blank lines to separate logical blocks of code, or control flow statements, unless they are the only ones within their parent block, in which case indentation suffices. You can look at the rest of the code for examples.

File: src/functions.js [hunk 9]
```diff
@@ -324,10 +324,48 @@ var _ = Mavo.Functions = {
...
+    },
+
 	/**
 	 * Case insensitive search
 	 */
-	search: (haystack, needle) => haystack && needle? str(haystack).toLowerCase().indexOf((needle + "").toLowerCase()) : -1,
+	search: (haystack, needle) => {
+		haystack = str(haystack);
+		needle = str(needle);
+		return haystack && needle? haystack.toLowerCase().indexOf((needle).toLowerCase()) : -1;
```
Comment (LeaVerou) [inline src/functions.js]: You don't need the parens around needle here.

**Comment by LeaVerou** [comment]
Huh, I thought I had made these comments before but I couldn't find them. Given that you just addressed them before I even posted them, I assume I had actually posted them before. Are there two open PRs about the same changes?!

**Comment by efeichen** [comment]
Your comment on `type` came in a few minutes ago, which made me remember that I only revised contains() in the other PR (the one with lots of discussions on booleans) and not this, so I changed all of them before your comments came.

And yes, PR #358 is branched off of this branch, so that's why that PR has so many commits in the beginning I think

**Comment by efeichen** [comment]
No, remember you told me to make a separate PR to change the binaryOperation() function in mavoscript.js? That's what #358 is for. The thing is I think it contains changes made by this PR's branch too because I branched that PR's branch off of this one.

Anyway, if everything looks good, I think merging the #358 will also add contains() function to functions.js, so I will close this one

**Pull request closed without merging**